File: src/CarGame.ts
import { Player, DIRECTIONS, Direction } from './Player.js';
import { TreesManager } from './TreeManager.js';
import { RoadManager, Road } from './RoadManager.js';
import { VehiclesManager } from './VehiclesManager.js';
import { Vehicle, VehicleCollisionObject } from './Vehicle.js';
import { EnvironmentObject } from './EnvironmentObjects.js';
import { Tree } from './Tree.js';


export interface Position {
    x: number;
    y: number;
}


// GameState class - manages the game state
export class GameState {
    // Constants =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
    //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

    player: Player;
    trees: Tree[];
    vehicles: Vehicle[];
    road: Road;
    gameOver: boolean;
    score: number;


    constructor() {
        this.player = new Player(0, CarGame.startY);
        this.road = []
        this.trees = [];
        this.vehicles = [];
        this.gameOver = false;
        this.score = 0;
    }

}

// Initialise the game-state
export class CarGame {
    // Constants =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
    readonly FPS: number = 120;
    readonly runLoopIntervalMilliseconds: number = 1000 / this.FPS / 2; // Twice the FPS
    // x -axis runs from -10 to 10
    readonly xAxisRange: number = 20;
    readonly roadWidth: number = this.xAxisRange / 2;
    static readonly startY: number = 10;

    //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

    gameState!: GameState;
    treesManager!: TreesManager;
    roadManager!: RoadManager;
    vehiclesManager!: VehiclesManager;

    //-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
    private runLoopInterval!: ReturnType<typeof setInterval>;

    constructor() {
        console.log('Welcome to Car Game 🚗');
        // 1.
        this.setupGameState();

        // 2
        this.setupKeyboardControls();

        //3.
        this.setupGameRunLoop();
    }

    setupGameState(): void {
        this.gameState = new GameState();

        this.roadManager = new RoadManager(this.gameState, this.roadWidth);
        this.treesManager = new TreesManager(this.gameState);
        this.vehiclesManager = new VehiclesManager(this.gameState);

        this.updateMap(); // trees and road are updated here    

    }



    setupKeyboardControls(): void {
        // Debounced / throttled keyboard controls
        document.addEventListener('keydown', (event: KeyboardEvent) => {
            // console.log('⌨️ keydown', event.key);
            switch (event.key) {
                case 'ArrowUp':
                case 'w':
                    this.gameState.player.adjustVelocity(DIRECTIONS.UP);
                    break;
                case 'ArrowDown':
                case 's':
                    this.gameState.player.adjustVelocity(DIRECTIONS.DOWN);
                    break;
                case 'ArrowLeft':
                case 'a':
                    this.gameState.player.adjustSteering(DIRECTIONS.LEFT);
                    break;
                case 'ArrowRight':
                case 'd':
                    this.gameState.player.adjustSteering(DIRECTIONS.RIGHT);
                    break;
            }
        });

        // TBC - touch controls -- can migrate from pacman later
        // document.addEventListener('touchstart', e => {
        //     this.touchstartX = e.changedTouches[0].screenX
        //     this.touchstartY = e.changedTouches[0].screenY
        // })

        // document.addEventListener('touchend', e => {
        //     this.touchendX = e.changedTouches[0].screenX
        //     this.touchendY = e.changedTouches[0].screenY
        //     this.checkSwipeDirection()
        // })
    }

    setupGameRunLoop(): void {
        this.runLoopInterval = setInterval(() => {
            this.gameRunLoop(this.runLoopIntervalMilliseconds);
        }, this.runLoopIntervalMilliseconds);
    }


    gameRunLoop(runLoopIntervalMilliseconds: number): void {
        this.updatePlayer(runLoopIntervalMilliseconds);
        this.updateMap(runLoopIntervalMilliseconds); // Update any foliage, enemies, road etc 


        this.checkAllCollisions();

        this.updateScore(); // do this last - after collisions are checked
    }

    updatePlayer(runLoopIntervalMilliseconds: number): void {
        this.gameState.player.updatePosition(runLoopIntervalMilliseconds / 1000);
    }

    updateMap(runLoopIntervalMilliseconds: number = 0): void {
        this.roadManager.updateRoad();
        this.vehiclesManager.updateVehicles(runLoopIntervalMilliseconds / 1000);

        this.treesManager.updateTrees(); // this requires a road to be present -- so it can put trees either side of the road

        // TBC - update other map elements
        // e.g. vehicles on the road
    }


    updateScore(): void {
        let currentY = Math.floor(this.gameState.player.y - CarGame.startY);
        // currentY = Math.floor(currentY / 10); // divide by 10 to scale the score down a bit.

        this.gameState.score = Math.max(this.gameState.score, currentY);
        // console.log('🚗🚗 Score: ', this.gameState.score);
    }

    checkAllCollisions(): void {

        // Check for collisions with the road and trees
        const objectListsToCheck = [this.gameState.trees, this.gameState.vehicles];


        for (let objectList of objectListsToCheck) {

            // Check for collisions in the list of objects
            const anyCollidedObjects = this.checkCollisionsInList(this.gameState.player, objectList);

            if (anyCollidedObjects) {
                this.endGame();
                return
            }
        }

    }

    checkCollisionsInList(player: Player, objectsToCheck: EnvironmentObject[]): boolean {

        // Player Max Size is used to quickly calculate if an object is even close to player
        const playerMaxSize = Math.max(player.width, player.length) * 1.5; // add buffer to the player's size

        const closeObjects = objectsToCheck.filter(
            object => object.checkObjectIsCloseToPlayer(player, playerMaxSize)
        );

        if (closeObjects.length > 0) {
            console.log('\n\n🌳🚗 Checking closeObjects for collisions', closeObjects.length, "/", objectsToCheck.length);

            // Get collision object for the player
            let playerCollissionObject: VehicleCollisionObject = player.getCollisionObject();
            // this is a snapshot of the player's position and orientation at the time of the check
            // with vertices and axes calculated for fast collision detection using SAT
            // Calculate this once before the for loop -- and only calculate it if there are closeObjects to check


            for (let object of closeObjects) {

                // We get a collision object, it's a snapshot of the object
                //  but in a representation that is easier to check for collisions
                let objectHasCollided = object.getCollisionObject().checkCollisionWithPlayerDetailed(playerCollissionObject)

                // For Trees the collisionObject is just the base object
                // For Vehicles the collisionObject is a VehicleCollisionObject with vertices and axes calculated for fast collision detection using SAT
                // Both use SAT but Trees calculate it based on a circle and the Vehicle calculates it based on a rotated rectangle
                // Trees could be simplified as a lot of the same calculations are repeated for each tree


                if (objectHasCollided) {
                    console.log('❌❌❌Object has collided -- gamer over ❌', object, player);
                    return true
                }
            }
        }

        return false

    }


    endGame() {
        console.log('-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=🚗🚗 Game Over 🚗🚗  -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=');
        clearInterval(this.runLoopInterval) // Cancel the run loop - stop it from updating
        this.gameState.gameOver = true;
    }



}

File: swift/CarDriveMAIN.playground/Sources/CarGame.swift
import Foundation
import SwiftUI

// MARK: - Core types

public typealias Road = (Position, Position)

public struct Position {
    public var x: Double
    public var y: Double

    public init(x: Double, y: Double) {
        self.x = x
        self.y = y
    }
}

// MARK: - GameState

public class GameState {
    public var player: Player
    public var trees: [Tree]
    public var vehicles: [Vehicle]
    public var road: [Road]
    public var gameOver: Bool
    public var score: Double

    public init() {
        self.player   = Player(x: 0, y: CarGame.startY)
        self.road     = []
        self.trees    = []
        self.vehicles = []
        self.gameOver = false
        self.score    = 0
    }
}

// MARK: - CarGame

public class CarGame {
    // World constants
    public static let xAxisRange: Double = 20.0
    public static let startY: Double     = 10.0
    let roadWidth: Double = CarGame.xAxisRange / 2.0

    public var gameState: GameState

    // Environment managers
    var roadManager: RoadManager!
    var treesManager: TreesManager!
    var vehiclesManager: VehiclesManager!

    public init() {
        print("Welcome to Car Game 🚗")
        gameState    = GameState()
        roadManager  = RoadManager(gameState, roadWidth)
        treesManager = TreesManager(gameState)
        vehiclesManager = VehiclesManager(gameState)
        updateMap()     // seed initial road, trees, vehicles
    }

    // MARK: - Game loop (called every frame by the view)

    public func gameRunLoop(dt: Double) {
        guard !gameState.gameOver else { return }
        updatePlayer(seconds: dt)
        updateMap(seconds: dt)
        checkAllCollisions()
        updateScore()
    }

    // MARK: - Updates

    func updatePlayer(seconds: Double) {
        gameState.player.updatePosition(seconds: seconds)
    }

    func updateMap(seconds: Double = 0) {
        roadManager.updateRoad()
        vehiclesManager.updateVehicles(seconds: seconds)
        treesManager.updateTrees()
    }

    func updateScore() {
        let dist = floor(gameState.player.y - CarGame.startY)
        gameState.score = max(gameState.score, dist)
    }

    // MARK: - Collision detection

    func checkAllCollisions() {
        if checkCollisionsInList(player: gameState.player, objects: gameState.trees as [EnvironmentObject]) {
            endGame(); return
        }
        if checkCollisionsInList(player: gameState.player, objects: gameState.vehicles as [EnvironmentObject]) {
            endGame(); return
        }
    }

    func checkCollisionsInList(player: Player, objects: [EnvironmentObject]) -> Bool {
        let maxSize = max(player.width, player.length) * 1.5
        let close   = objects.filter { $0.checkObjectIsCloseToPlayer(player, playerMaxSize: maxSize) }
        guard !close.isEmpty else { return false }

        let playerCollision = player.getVehicleCollisionObject()
        for obj in close {
            if obj.getCollisionObject().checkCollisionWithPlayerDetailed(playerCollision) {
                return true
            }
        }
        return false
    }

    // MARK: - Game over / restart

    func endGame() {
        print("🚗🚗 Game Over! Score: \(Int(gameState.score)) 🚗🚗")
        gameState.gameOver = true
    }

    public func restart() {
        gameState       = GameState()
        roadManager     = RoadManager(gameState, roadWidth)
        treesManager    = TreesManager(gameState)
        vehiclesManager = VehiclesManager(gameState)
        updateMap()
    }
}
